src/System.Private.CoreLib/shared/System/MemoryExtensions.cs

```diff
@@ -1606,5 +1606,111 @@ public static bool Overlaps<T>(this ReadOnlySpan<T> span, ReadOnlySpan<T> other,
+        /// using the <typeparamref name="TComparer" />.
+        /// </summary>
+        public static void Sort<T, TComparer>(this Span<T> span, TComparer comparer)
+            where TComparer : IComparer<T>
+        {
+            if (span.Length > 1)
+            {
+                // Span based Sort does not call `TrySZSort`, but instead
+                // uses the managed generic code only. This means Span based Sort 
+                // might differ in both performance and sorting for equal keys
+
+                // TODO: Add TComparer support until then value type comparers will be boxed
+                //       and performance will be affected
+                ArraySortHelper<T>.Default.Sort(span, 0, span.Length, comparer);
+            }
+        }
+
+        /// <summary>
+        /// Sorts the elements in the entire <see cref="Span{T}" /> 
+        /// using the <see cref="Comparison{T}" />.
+        /// </summary>
+        public static void Sort<T>(this Span<T> span, Comparison<T> comparison)
+        {
+            if (comparison == null)
+                ThrowHelper.ThrowArgumentNullException(ExceptionArgument.comparison);
+            if (span.Length > 1)
+            {
+                ArraySortHelper<T>.Sort(span, 0, span.Length, comparison!); // TODO-NULLABLE: https://github.com/dotnet/csharplang/issues/538
+            }
+        }
+
+        /// <summary>
+        /// Sorts a pair of spans 
+        /// (one contains the keys <see cref="Span{TKey}"/> 
+        /// and the other contains the corresponding items <see cref="Span{TValue}"/>) 
+        /// based on the keys in the first <see cref= "Span{TKey}" /> 
+        /// using the <see cref="IComparable" /> implementation of each 
+        /// element of the <see cref= "Span{TKey}"/>.
+        /// </summary>
+        public static void Sort<TKey, TValue>(this Span<TKey> keys, Span<TValue> items) =>
+            Sort(keys, items, (IComparer<TKey>)null!);
+
+        /// <summary>
+        /// Sorts a pair of spans 
+        /// (one contains the keys <see cref="Span{TKey}"/> 
+        /// and the other contains the corresponding items <see cref="Span{TValue}"/>) 
+        /// based on the keys in the first <see cref= "Span{TKey}" /> 
+        /// using the <typeparamref name="TComparer" />.
+        /// </summary>
+        public static void Sort<TKey, TValue, TComparer>(this Span<TKey> keys,
+            Span<TValue> items, TComparer comparer)
+            where TComparer : IComparer<TKey>
+        {
+            if (keys.Length != items.Length)
+                ThrowHelper.ThrowArgumentException(ExceptionResource.Argument_SpansMustHaveSameLength);
+
+            if (keys.Length > 1)
+            {
+                // Span based Sort does not call `TrySZSort`, but instead
+                // uses the managed generic code only. This means Span based Sort 
+                // might differ in both performance and sorting for equal keys
```

Comment:
It shouldn't affect this PR, but have you measure to see what the difference is between an array-based sort that does use TrySZSort and a span-based sort that doesn't?  How far off are they?

Reply:
@stephentoub comparing `int` vs `IntStruct` can indicate this (showing the reverse in this case).

# `int`
|               Method |     Toolchain | Size |      Mean |     Error |    StdDev |    Median |       Min |       Max | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|--------------------- |-------------- |----- |----------:|----------:|----------:|----------:|----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
|                Array | \coreclr-base |  512 |  5.665 us | 0.0367 us | 0.0343 us |  5.670 us |  5.608 us |  5.725 us |  1.00 |    0.00 |           - |           - |           - |                   - |
|                Array |      \coreclr |  512 |  5.697 us | 0.0393 us | 0.0368 us |  5.693 us |  5.636 us |  5.779 us |  1.01 |    0.01 |           - |           - |           - |                   - |


# `IntStruct`
|               Method |      Toolchain | Size |      Mean |     Error |    StdDev |    Median |       Min |       Max | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|--------------------- |--------------- |----- |----------:|----------:|----------:|----------:|----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
|                Array | \coreclr-base\ |  512 |  3.785 us | 0.1062 us | 0.1222 us |  3.714 us |  3.684 us |  4.021 us |  1.00 |    0.00 |           - |           - |           - |                   - |
|                Array |      \coreclr\ |  512 |  3.641 us | 0.0299 us | 0.0280 us |  3.646 us |  3.581 us |  3.701 us |  0.96 |    0.03 |           - |           - |           - |                   - |

Quite a difference here in favor of the managed code path.

Note that this is not just about perf, native sorting code does a `NaNPrepass` (if I remember correctly) for floats, that the span version would then not do.

Reply:
Above is for

``` ini

BenchmarkDotNet=v0.11.3.1003-nightly, OS=Windows 10.0.17763.503 (1809/October2018Update/Redstone5)
Intel Core i7-8700 CPU 3.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.0.100-preview5-011568
  [Host]     : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT
  Job-JTKJDV : .NET Core ? (CoreCLR 4.6.27803.0, CoreFX 4.700.19.22901), 64bit RyuJIT
  Job-DBBNCI : .NET Core ? (CoreCLR 4.6.27808.0, CoreFX 4.700.19.22901), 64bit RyuJIT

Runtime=Core  InvocationCount=10000  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  UnrollFactor=1  
WarmupCount=1  

```

Reply:
@stephentoub you can see some more benchmarks in the gist here:

https://gist.github.com/nietras/ccf8199a9722a396731a0597d55916bf

Reply:
@stephentoub after merging with master I have rerun the benchmarks with higher invocation count (25000). This is more inline with what I expected with regards to native (`int`) vs managed code (`IntStruct`). native being a bit faster with the caveats that this is based on using `IntStruct` as a replacement for `int`.

coreclr-base = 5af77fa0dcc99abb3c9610f1a062493ddc2e0584 (master)
coreclr = 11dd34a67f2a78b46dabc921e32ba63731057d63 (from this PR branch)

# Int32
``` ini

BenchmarkDotNet=v0.11.3.1003-nightly, OS=Windows 10.0.17763.503 (1809/October2018Update/Redstone5)
Intel Core i7-8700 CPU 3.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.0.100-preview5-011568
  [Host]     : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT
  Job-VKBMMT : .NET Core ? (CoreCLR 4.700.19.30301, CoreFX 4.700.19.27908), 64bit RyuJIT
  Job-SABOCD : .NET Core ? (CoreCLR 4.700.19.30301, CoreFX 4.700.19.27908), 64bit RyuJIT

Runtime=Core  InvocationCount=25000  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  UnrollFactor=1  
WarmupCount=1  

```
|               Method |               | Size |      Mean |     Error |    StdDev |    Median |       Min |       Max | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|--------------------- |-------------- |----- |----------:|----------:|----------:|----------:|----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
|                Array | \coreclr-base |  512 |  3.495 us | 0.1296 us | 0.1387 us |  3.469 us |  3.323 us |  3.753 us |  1.00 |    0.00 |           - |           - |           - |                   - |
|                Array |      \coreclr |  512 |  3.640 us | 0.2213 us | 0.2460 us |  3.573 us |  3.317 us |  4.222 us |  1.05 |    0.08 |           - |           - |           - |                   - |
|                      |               |      |           |           |           |           |           |           |       |         |             |             |             |                     |
|  Array_ComparerClass | \coreclr-base |  512 | 19.367 us | 0.1069 us | 0.0948 us | 19.357 us | 19.207 us | 19.560 us |  1.00 |    0.00 |           - |           - |           - |                64 B |
|  Array_ComparerClass |      \coreclr |  512 | 19.474 us | 0.1196 us | 0.1060 us | 19.452 us | 19.321 us | 19.695 us |  1.01 |    0.01 |           - |           - |           - |                64 B |
|                      |               |      |           |           |           |           |           |           |       |         |             |             |             |                     |
| Array_ComparerStruct | \coreclr-base |  512 | 22.870 us | 0.1593 us | 0.1490 us | 22.820 us | 22.696 us | 23.192 us |  1.00 |    0.00 |           - |           - |           - |                88 B |
| Array_ComparerStruct |      \coreclr |  512 | 23.718 us | 0.0634 us | 0.0593 us | 23.709 us | 23.614 us | 23.859 us |  1.04 |    0.01 |           - |           - |           - |                88 B |
|                      |               |      |           |           |           |           |           |           |       |         |             |             |             |                     |
|     Array_Comparison | \coreclr-base |  512 | 18.615 us | 0.0971 us | 0.0861 us | 18.598 us | 18.519 us | 18.826 us |  1.00 |    0.00 |           - |           - |           - |                   - |
|     Array_Comparison |      \coreclr |  512 | 19.760 us | 0.2582 us | 0.2415 us | 19.647 us | 19.556 us | 20.369 us |  1.06 |    0.01 |           - |           - |           - |                   - |
|                      |               |      |           |           |           |           |           |           |       |         |             |             |             |                     |
|                 List | \coreclr-base |  512 |  3.385 us | 0.0916 us | 0.0980 us |  3.335 us |  3.298 us |  3.607 us |  1.00 |    0.00 |           - |           - |           - |                   - |
|                 List |      \coreclr |  512 |  3.759 us | 0.4917 us | 0.5465 us |  3.534 us |  3.279 us |  5.224 us |  1.10 |    0.17 |           - |           - |           - |                   - |
|                      |               |      |           |           |           |           |           |           |       |         |             |             |             |                     |

# IntStruct
``` ini

BenchmarkDotNet=v0.11.3.1003-nightly, OS=Windows 10.0.17763.503 (1809/October2018Update/Redstone5)
Intel Core i7-8700 CPU 3.20GHz (Coffee Lake), 1 CPU, 12 logical and 6 physical cores
.NET Core SDK=3.0.100-preview5-011568
  [Host]     : .NET Core 3.0.0-preview5-27626-15 (CoreCLR 4.6.27622.75, CoreFX 4.700.19.22408), 64bit RyuJIT
  Job-VKBMMT : .NET Core ? (CoreCLR 4.700.19.30301, CoreFX 4.700.19.27908), 64bit RyuJIT
  Job-SABOCD : .NET Core ? (CoreCLR 4.700.19.30301, CoreFX 4.700.19.27908), 64bit RyuJIT

Runtime=Core  InvocationCount=25000  IterationTime=250.0000 ms  
MaxIterationCount=20  MinIterationCount=15  UnrollFactor=1  
WarmupCount=1  

```
|               Method |               | Size |      Mean |     Error |    StdDev |    Median |       Min |       Max | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|--------------------- |-------------- |----- |----------:|----------:|----------:|----------:|----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
|                Array | \coreclr-base |  512 |  3.729 us | 0.0498 us | 0.0416 us |  3.731 us |  3.635 us |  3.786 us |  1.00 |    0.00 |           - |           - |           - |                   - |
|                Array |      \coreclr |  512 |  3.925 us | 0.2357 us | 0.2522 us |  3.913 us |  3.654 us |  4.505 us |  1.07 |    0.08 |           - |           - |           - |                   - |
|                      |               |      |           |           |           |           |           |           |       |         |             |             |             |                     |
|  Array_ComparerClass | \coreclr-base |  512 | 20.788 us | 0.0535 us | 0.0474 us | 20.782 us | 20.669 us | 20.867 us |  1.00 |    0.00 |           - |           - |           - |                64 B |
|  Array_ComparerClass |      \coreclr |  512 | 21.212 us | 0.0889 us | 0.0788 us | 21.192 us | 21.101 us | 21.398 us |  1.02 |    0.00 |           - |           - |           - |                64 B |
|                      |               |      |           |           |           |           |           |           |       |         |             |             |             |                     |
| Array_ComparerStruct | \coreclr-base |  512 | 23.963 us | 0.0966 us | 0.0856 us | 23.949 us | 23.830 us | 24.183 us |  1.00 |    0.00 |           - |           - |           - |                88 B |
| Array_ComparerStruct |      \coreclr |  512 | 22.890 us | 0.1253 us | 0.1047 us | 22.894 us | 22.684 us | 23.034 us |  0.96 |    0.01 |           - |           - |           - |                88 B |
|                      |               |      |           |           |           |           |           |           |       |         |             |             |             |                     |
|     Array_Comparison | \coreclr-base |  512 | 20.161 us | 0.1313 us | 0.1164 us | 20.157 us | 19.995 us | 20.401 us |  1.00 |    0.00 |           - |           - |           - |                   - |
|     Array_Comparison |      \coreclr |  512 | 21.104 us | 0.0972 us | 0.0910 us | 21.101 us | 20.995 us | 21.292 us |  1.05 |    0.01 |           - |           - |           - |                   - |
|                      |               |      |           |           |           |           |           |           |       |         |             |             |             |                     |
|                 List | \coreclr-base |  512 |  4.891 us | 1.0011 us | 1.1528 us |  4.679 us |  3.741 us |  7.906 us |  1.00 |    0.00 |           - |           - |           - |                   - |
|                 List |      \coreclr |  512 |  4.047 us | 0.1858 us | 0.1988 us |  4.076 us |  3.765 us |  4.503 us |  0.84 |    0.17 |           - |           - |           - |                   - |